applications/GeoMechanicsApplication/custom_utilities/nodal_extrapolator.cpp

```diff
@@ -0,0 +1,108 @@
+// KRATOS___
+//     //   ) )
+//    //         ___      ___
+//   //  ____  //___) ) //   ) )
+//  //    / / //       //   / /
+// ((____/ / ((____   ((___/ /  MECHANICS
+//
+//  License:         geo_mechanics_application/license.txt
+//
+//  Main authors:    Richard Faasse
+//
+
+#include "nodal_extrapolator.h"
+#include "geometries/quadrilateral_2d_4.h"
+#include "geometries/triangle_2d_3.h"
+
+namespace Kratos
+{
+
+Matrix NodalExtrapolator::CalculateElementExtrapolationMatrix(GeometryType& rGeometry,
+                                                              GeometryData::IntegrationMethod IntegrationMethod) const
+{
+    CheckIfGeometryIsSupported(rGeometry);
+
+    auto p_lower_order_geometry = CreateLowerOrderGeometry(rGeometry);
+    const GeometryType& p_corner_geometry = p_lower_order_geometry ? *p_lower_order_geometry : rGeometry;
+
+    Matrix extrapolation_matrix =
+        CalculateExtrapolationMatrixForCornerNodes(rGeometry, IntegrationMethod, p_corner_geometry);
+
+    if (p_lower_order_geometry) {
+        AddRowsForMidsideNodes(rGeometry, extrapolation_matrix);
+    }
+
+    return extrapolation_matrix;
+}
+
+Matrix NodalExtrapolator::CalculateExtrapolationMatrixForCornerNodes(
+    const NodalExtrapolator::GeometryType& rGeometry,
+    const GeometryData::IntegrationMethod& IntegrationMethod,
+    const NodalExtrapolator::GeometryType& rCornerGeometry) const
+{
+    SizeType number_of_corner_nodes = rCornerGeometry.PointsNumber();
+
+    Matrix      quasi_mass_mat     = ZeroMatrix(number_of_corner_nodes, number_of_corner_nodes);
+    const auto& integration_points = rGeometry.IntegrationPoints(IntegrationMethod);
+    const auto  number_of_integration_points = integration_points.size();
+    Matrix      node_coefficient(number_of_corner_nodes, number_of_integration_points);
```

Comment:
Since this is a matrix with values, I would use the plural form rather than the singular form:
```suggestion
    Matrix      node_coefficients(number_of_corner_nodes, number_of_integration_points);
```